### PR TITLE
implement double backwards for MaxPool3d

### LINKED ISSUE #4497: MaxPool3d cannot be differentiated twice
While trying to implement WGan with gradient penalty for 3D volumes, I realized that MaxPool3d cannot be differentiated twice, while it works for MaxPool2d or even AvgPool3d.

Here is the code I am using :

```python

class Network(nn.Module):
    def __init__(self):
        super().__init__()
        self.main = nn.AvgPool3d(2) # Works
        # self.main = nn.MaxPool3d(2) # Does not work
        # self.main = nn.MaxPool2d(2) # Works for 2d

    def forward(self, x):
        return self.main(x)

net = Network()
net.cuda()
in_var = Variable(torch.Tensor(2, 2, 2, 2, 2).cuda(), requires_grad=True)
grad_out = Variable(torch.ones(2, 2, 1, 1, 1).cuda())

# in_var = Variable(torch.Tensor(2, 2, 2, 2).cuda(), requires_grad=True) # Works for 2d
# grad_out = Variable(torch.ones(2, 2, 1, 1).cuda()) # Works for 2d


gradient = autograd.grad(outputs=net(in_var), inputs=in_var,
                         grad_outputs=grad_out,
                         create_graph=True, retain_graph=True,
                         only_inputs=True)[0]
gradient.mean().backward()

```
Thanks

### PR DESCRIPTION
**Pull request by li-roy**
fixes #4497.

test plan: added double backwards check for maxpool3d tests

### CODE REVIEW

**Comment by zou3519** [comment]
Could you run some benchmarks to see if the memory usage increase from changing MaxUnpool3d to not pack indices is significant? I think someone (@apaszke) mentioned Resnet as a network that uses MaxUnpool3d; it would be nice to see if the memory usage increases significantly or not under resnet.

**Comment by apaszke** [comment]
Oh, resnets use the 2d version, and I thought we’re modifying that. Will this make 3d consistent with the other pooling layers?

**Comment by li-roy** [comment]
Yeah this makes it more consistent. Also, I don't see any reason for this to use more memory than the old implementation. The way indices are being stored is changed, but we're still using the same number of bits in the tensor.

File: aten/src/THNN/generic/VolumetricMaxUnpooling.c [hunk 1]
```diff
@@ -63,49 +63,36 @@ static void THNN_(VolumetricMaxUnpooling_updateOutput_frame)(
...
-          THIndex_t *ind_p_k = ind_p + k*iT*iW*iH + ti*iW*iH + i*iW + j;
-
-          maxz = ((unsigned char*)(ind_p_k))[0]; /* retrieve position of max */
-          maxy = ((unsigned char*)(ind_p_k))[1];
-          maxx = ((unsigned char*)(ind_p_k))[2];
-
-          THIndex_t idx = k*oT*oW*oH + oH*oW*(start_t+maxz) + oW*(start_h+maxy) + (start_w+maxx);
-          if (start_t+maxz<0 || start_h+maxy<0 || start_w+maxx<0 || start_t+maxz>=oT
-	      || start_h+maxy>=oH || start_w+maxx>=oW)
+          maxp = ind_p_k[ti * iH * iW + i * iW + j] - TH_INDEX_BASE;  /* retrieve position of max */
```
Comment (zou3519) [inline aten/src/THNN/generic/VolumetricMaxUnpooling.c]: Nit: put ti*iH*iW + i * iW + j into a variable and reuse below on line 95

File: aten/src/THNN/generic/VolumetricMaxUnpooling.c [hunk 2]
```diff
@@ -63,49 +63,36 @@ static void THNN_(VolumetricMaxUnpooling_updateOutput_frame)(
...
   for (k = 0; k < nslices; k++)
   {
-    int ti, i, j, maxz, maxy, maxx;
+    real *output_p_k = output_p + k * oT * oH * oW;
+    real *input_p_k = input_p + k * iT * iH * iW;
+    THIndex_t *ind_p_k = ind_p + k * iT * iH * iW;
+
+    int ti, i, j;
+    THIndex_t maxp;
     for (ti = 0; ti < iT; ti++)
```
Comment (zou3519) [inline aten/src/THNN/generic/VolumetricMaxUnpooling.c]: Nit: this is probably nicer as (t, h, w) instead of (ti, i, j).
Same for the usage in maxpool3d.

File: aten/src/THNN/generic/VolumetricMaxUnpooling.c [hunk 3]
```diff
@@ -228,46 +212,30 @@ static void THNN_(VolumetricMaxUnpooling_updateGradInput_frame)(
...
   for (k = 0; k < nslices; k++)
   {
-    int ti, i, j, maxz, maxy, maxx;
+    real *gradInput_p_k = gradInput_p + k * iT * iH * iW;
+    real *gradOutput_p_k = gradOutput_p + k * oT * oH * oW;
+    THIndex_t *ind_p_k = ind_p + k * iT * iH * iW;
+
+    int ti, i, j;
+    THIndex_t maxp;
     for (ti = 0; ti < iT; ti++)
```
Comment (zou3519) [inline aten/src/THNN/generic/VolumetricMaxUnpooling.c]: Nit: replace loop variables with (t, h, w)?

File: aten/src/THNN/generic/VolumetricDilatedMaxPooling.c [hunk 4]
```diff
@@ -381,16 +372,13 @@ static void THNN_(VolumetricDilatedMaxPooling_updateGradInput_frame)(
...
-	    /* update gradient */
-	    gradInput_p_k[maxti * iheight * iwidth + maxi * iwidth + maxj] +=
-	      gradOutput_p_k[ti * oheight * owidth + i * owidth + j];
-	  }
+          int64_t index = ti * oheight * owidth + i * owidth + j;
+          int64_t maxp = indz_p_k[index] - TH_INDEX_BASE;
+
+          if (maxp != -1) {
+            /* update gradient */
+            gradInput_p_k[maxp] += gradOutput_p_k[index];
```
Comment (zou3519) [inline aten/src/THNN/generic/VolumetricDilatedMaxPooling.c]: Is it sufficient to just assign gradInput_p_k[maxp] = gradOutput_p_k[index]?

**Review by zou3519 — COMMENTED** [review]
The math looks good to me. I had some minor comments.

File: aten/src/THNN/generic/VolumetricDilatedMaxPooling.c (hunk 4 above)
Comment (li-roy) [reply aten/src/THNN/generic/VolumetricDilatedMaxPooling.c]: Nah, different outputs can depend on the same input.

File: aten/src/THNN/generic/VolumetricDilatedMaxPooling.c [hunk 5]
```diff
@@ -127,6 +127,7 @@ static void THNN_(VolumetricDilatedMaxPooling_updateOutput_frame)(
   {
     /* loop over output */
     int64_t i, j, ti;
+    real *ip = input_p + k * itime * iwidth * iheight;
     for (ti = 0; ti < otime; ti++)
```
Comment (zou3519) [inline aten/src/THNN/generic/VolumetricDilatedMaxPooling.c]: nit: ti -> ot, i -> oh, j -> ow?

File: aten/src/THNN/generic/VolumetricDilatedMaxPooling.c [hunk 6]
```diff
@@ -150,46 +151,36 @@ static void THNN_(VolumetricDilatedMaxPooling_updateOutput_frame)(
...
+          int64_t index = 0;
 
-          for (z = 0; z < kernel_t; z++)
+          for (z = start_t; z < end_t; z += dilationT)
           {
-            for (y = 0; y < kernel_h; y++)
+            for (y = start_h; y < end_h; y += dilationH)
             {
-              for (x = 0; x < kernel_w; x++)
+              for (x = start_w; x < end_w; x += dilationW)
```
Comment (zou3519) [inline aten/src/THNN/generic/VolumetricDilatedMaxPooling.c]: nit: z -> t, y -> h, x -> w

File: aten/src/THNN/generic/VolumetricDilatedMaxPooling.c [hunk 7]
```diff
@@ -381,16 +372,13 @@ static void THNN_(VolumetricDilatedMaxPooling_updateGradInput_frame)(
         for (j = 0; j < owidth; j++)
```
Comment (zou3519) [inline aten/src/THNN/generic/VolumetricDilatedMaxPooling.c]: nit: ti -> t, i -> h, j -> w

File: aten/src/THNN/generic/VolumetricMaxUnpooling.c [hunk 8]
```diff
@@ -63,49 +63,37 @@ static void THNN_(VolumetricMaxUnpooling_updateOutput_frame)(
...
+    real *input_p_k = input_p + k * iT * iH * iW;
+    THIndex_t *ind_p_k = ind_p + k * iT * iH * iW;
+
+    int t, i, j, index;
+    THIndex_t maxp;
+    for (t = 0; t < iT; t++)
     {
       for (i = 0; i < iH; i++)
       {
         for (j = 0; j < iW; j++)
```
Comment (zou3519) [inline aten/src/THNN/generic/VolumetricMaxUnpooling.c]: nit: i -> h, j -> w

File: aten/src/THNN/generic/VolumetricMaxUnpooling.c [hunk 9]
```diff
@@ -228,46 +213,31 @@ static void THNN_(VolumetricMaxUnpooling_updateGradInput_frame)(
...
+    real *gradOutput_p_k = gradOutput_p + k * oT * oH * oW;
+    THIndex_t *ind_p_k = ind_p + k * iT * iH * iW;
+
+    int t, i, j, index;
+    THIndex_t maxp;
+    for (t = 0; t < iT; t++)
     {
       for (i = 0; i < iH; i++)
       {
         for (j = 0; j < iW; j++)
```
Comment (zou3519) [inline aten/src/THNN/generic/VolumetricMaxUnpooling.c]: nit: i -> h, j -> w

**Review by zou3519 — COMMENTED** [review]
More nits about the naming that existed previously in the code (before your changes). Feel free to ignore

**Comment by zou3519** [comment]
This will also fix #1197 I believe, thanks @li-roy :D

**Review by zou3519 — APPROVED** [review]
LGTM!
